### PR TITLE
Scannet

### PR DESCRIPTION
**Pull request by Chidubem17**

### CODE REVIEW

File: data_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
+import tensorflow as tf
+import numpy as np
+from projection_utils import get_axismatrix, projection_to_rgb
+import os
```
Comment (ahmadyan) [inline data_pipeline.py]: import generic stuff, like os, first. Then thirdparty, like numpy, finally your own stuff.

File: data_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+        depths: array of depth images for each scene
+        extrinsic_matrix: (4x4 array): camera to world extrinsic parameters
+        intrinsic_matrix: (4x4 array): camera intrinsic parameters
+        axis_alinment: (4x4 array): axis_alignment matrix
+        box_objects: Nx7 array containing box center, size and angle of N boxes
+    Return: serialized tensorflow example of a scene
+    """
+
+    feature = {
+        'images': _bytes_feature(images),
```
Comment (ahmadyan) [inline data_pipeline.py:53]: Add a comment, if you expect this to contains multiple images. Also specify what is the type of the image (uint8 opencv or numpy array)

File: data_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+        'depths': _bytes_feature(depths),
+        'extrinsic_matrix': _bytes_feature(extrinsic_matrix),
+        'intrinsic_matrix': _bytes_feature(intrinsic_matrix),
+        'axis_matrix': _bytes_feature(axis_matrix),
+        'box_objects': _bytes_feature(box_objects),
+    }
+
+    #  Create a Features message using tf.train.Example.
```
Comment (ahmadyan) [inline data_pipeline.py]: remove space, and "Feature"

File: projection_utils.py [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+    ax.w_zaxis.set_pane_color((0.8, 0.8, 0.8, 1.0))
+
+    # rotate the axes and update
+    ax.view_init(30, 12)
+    plt.draw()
+    plt.savefig(name)
+    plt.show()
+
+
+def draw_2dboxes(imgname, boxes=[], points=True, greyscale=False, flip=True):
```
Comment (ahmadyan) [inline projection_utils.py]: pick a better name than imgname

File: projection_utils.py [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+        plydata = PlyData.read(f)
+        num_verts = plydata['vertex'].count
+        vertices = np.zeros(shape=[num_verts, 3], dtype=np.float32)
+        vertices[:, 0] = plydata['vertex'].data['x']
+        vertices[:, 1] = plydata['vertex'].data['y']
+        vertices[:, 2] = plydata['vertex'].data['z']
+    return vertices
+
+
+def alignaxis(axisfile, plyfile, sample=False):
```
Comment (ahmadyan) [inline projection_utils.py]: function name is not styled correctly, same for axisfile, plyfile,

Comment (ahmadyan) [reply projection_utils.py]: Add sample to the docstring.

File: data_pipeline.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+    depths = []
+    extrinsics = []
+    assert len(image_paths) == len(depth_paths) == len(extrinsic_matrix_paths)
+    with tf.io.TFRecordWriter(filename) as writer:
+        for image_path, extrinsic_path, depth_path in zip(
+                image_paths, extrinsic_matrix_paths, depth_paths):
+
+            img = tf.keras.preprocessing.image.load_img(image_path)
+            img_array = tf.keras.preprocessing.image.img_to_array(img)
+            img_array = img_array
```
Comment (ahmadyan) [inline data_pipeline.py]: what does this line do?

File: data_pipeline.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+            images.append(img_array)
+
+            img = tf.keras.preprocessing.image.load_img(depth_path)
+            img_array = tf.keras.preprocessing.image.img_to_array(img)
+            img_array = img_array
```
Comment (ahmadyan) [inline data_pipeline.py]: ditto

File: data_pipeline.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+        for image_path, extrinsic_path, depth_path in zip(
+                image_paths, extrinsic_matrix_paths, depth_paths):
+
+            img = tf.keras.preprocessing.image.load_img(image_path)
+            img_array = tf.keras.preprocessing.image.img_to_array(img)
+            img_array = img_array
+            images.append(img_array)
+
+            img = tf.keras.preprocessing.image.load_img(depth_path)
+            img_array = tf.keras.preprocessing.image.img_to_array(img)
```
Comment (ahmadyan) [inline data_pipeline.py]: rename to depth_image_array

File: data_pipeline.py [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+
+    scene = os.path.basename(scene_dir)
+    axis_matrix = get_axismatrix(os.path.join(scene_dir, scene + '.txt'))
+    intrinsic_matrix = np.loadtxt(
+        os.path.join(
+            scene_dir,
+            'intrinsic_color.txt'))
+    box_objects = np.loadtxt(os.path.join(scene_dir, 'box_coords.txt'))
+
+    # assuming all images have the same width and height, change if not true
```
Comment (ahmadyan) [inline data_pipeline.py]: Start your comment with the capital letters: Assuming, and end it with .
Also applied to the other comments in this PR as well.

File: data_pipeline.py [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+    height = 968
```
Comment (ahmadyan) [inline data_pipeline.py]: Move this stuff as global variables to the top of the file, or make them input.

File: data_pipeline.py [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+        count = 0
+        for i, b in enumerate(boxes):
+            b = np.float32(b)
+            x, y = b[:, 0], b[:, 1]
+            if (x[0] < width) and (y[0] < height):
+                count += 1
+        center_counts.append(count)
+
+    # sort in center counts in descending order and select the top 30 images
+    center_indexes = np.argsort(center_counts)[::-1][:30]
```
Comment (ahmadyan) [inline data_pipeline.py]: what if it is visible in less than 30 images?

File: data_pipeline.py [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+    assert center_indexes.size >= train_size, "reduce train size"
+    # randomly select train_size images from the list
+    np.random.seed(2000)
```
Comment (ahmadyan) [inline data_pipeline.py]: Use a singular magic seed number, at the begining of your main function. Not here.

File: data_pipeline.py [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+    depth_paths = [os.path.join(depth_paths, images[i][:-3] + 'png')
+                   for i in center_indexes]
+    extrinsic_paths = [os.path.join(
+        extrinsic_paths, images[i][:-3] + 'txt') for i in center_indexes]
+
+    return image_paths, depth_paths, extrinsic_paths
+
+
+def get_scene_data(scenename):
+    """Reads input data from scene
```
Comment (ahmadyan) [inline data_pipeline.py]: If you require any pre-processing to be done to the ScanNet prior to running this function, explain it here.

File: data_pipeline.py [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+    """Create input pipeline for the model
+    Args:
+        tfrecords: file lists or file of tfrecords
+    Return: dataset for training
+    """
+
+    dataset = tf.data.TFRecordDataset(tfrecords)
+    dataset = dataset.map(read_decode_tfrecords, num_parallel_calls=4)
+# dataset = dataset.shuffle(400).repeat().batch(BATCH_SIZE) # box objects
+# have different shape so cannot be batched
```
Comment (ahmadyan) [inline data_pipeline.py]: fix the style

File: projection_utils.py [hunk 15]
```diff
@@ -0,0 +1,324 @@
...
+    mesh_vertices[:, 0:3] = pts[:, 0:3]
+
+    if sample:
+        pts = np.arange(pts.shape[0])
+        pts = np.random.choice(pts, 5000, replace=False)
+
+    return mesh_vertices[pts]
+
+
+def rotationmat(heading_angle):
```
Comment (ahmadyan) [inline projection_utils.py]: fix function names

File: projection_utils.py [hunk 16]
```diff
@@ -0,0 +1,324 @@
...
+    rotmat = np.zeros((3, 3))
+    rotmat[2, 2] = 1
+    cosval = np.cos(heading_angle)
+    sinval = np.sin(heading_angle)
+    rotmat[0:2, 0:2] = np.array([[cosval, -sinval], [sinval, cosval]])
+    return rotmat
+
+
+def local_coord(scale):
+    """Define a 3d coord cube centered at 0
```
Comment (ahmadyan) [inline projection_utils.py]: 3D axis-aligned cube

File: projection_utils.py [hunk 17]
```diff
@@ -0,0 +1,324 @@
...
+    d = scale[2] / 2.
+
+    # Define the local coordinate system, w.r.t. the center of the box
+    box_coords = np.array([[0., 0., 0.], [-w, -h, -d], [-w, -h, +d], [-w, +h, -d],
+                           [-w, +h, +d], [+w, -h, -d], [+w, -h, +d], [+w, +h, -d],
+                           [+w, +h, +d]])
+    return box_coords
+
+
+def gen_vertices(box_points, heading_angle):
```
Comment (ahmadyan) [inline projection_utils.py:111]: doc string is missing heading_angle

File: projection_utils.py [hunk 18]
```diff
@@ -0,0 +1,324 @@
...
+    """Generate homogenous vertices of bounding boxes given center, size and heading angle of box
+        Args:
+          boxpoint: size 7 array containing box center, size and angle in degrees
+
+        Return: 9x4 array of vertices for bounding box
+    """
+    # convert angle to radians
+    heading_angle = heading_angle * np.pi / 180
+
+    ctr = box_points[:3]
```
Comment (ahmadyan) [inline projection_utils.py]: Use descriptive names when not short on space, like center, transformations, etc. makes reading your code a lot easier.

File: projection_utils.py [hunk 19]
```diff
@@ -0,0 +1,324 @@
...
+
+def create_aligned_pose(axis_alignment, extrinsic_matrix):
+    """Axis align the extrinsic matrix and convert to world to camera coords
+    Args:
+        axis_alinment: axis_alignment matrix
+        extrinsic_matrix: (4x4 array): camera to world extrinsic parameters
+    Return: axis aligned world to camera pose
+    """
+    pose = np.matmul(axis_alignment, extrinsic_matrix)
+    pose = np.linalg.inv(pose)
```
Comment (ahmadyan) [inline projection_utils.py:199]: Comment/Explain why you are inverting the matrix here.

**Comment by ahmadyan** [comment]
Overall it looks good, some minor comments mostly around styling.

**Comment by ahmadyan** [comment]
LGTM
